cpp/ql/src/experimental/Security/CWE/CWE-243/IncorrectChangingWorkingDirectory.ql

```diff
@@ -0,0 +1,70 @@
+/**
+ * @name Find work with changing working directories, with security errors.
+ * @description Not validating the return value or pinning the directory can be unsafe.
+ * @kind problem
+ * @id cpp/work-with-changing-working-directories
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-243
+ *       external/cwe/cwe-252
+ */
+
+import cpp
+
+/** Holds if a `fc` function call is available before or before a `chdir` function call. */
+predicate inExistsChdir(FunctionCall fcp) {
+  exists(FunctionCall fctmp |
+    (
+      fctmp.getTarget().hasGlobalOrStdName("chdir") or
+      fctmp.getTarget().hasGlobalOrStdName("fchdir")
+    ) and
+    (
+      fctmp.getASuccessor*() = fcp or
+      fcp.getASuccessor*() = fctmp
```

Comment:
Thanks.
it definitely improved detection.
removing cases when there was a call as an argument.

Reply:
Good afternoon @MathiasVP.

i changed these two lines (in two predicates) to
```ql
  fcp.getBasicBlock().getASuccessor*()=fctmp.getBasicBlock() or
  fctmp.getBasicBlock().getASuccessor*()=fcp.getBasicBlock()
```

and got additional detection in project `joncampbell123/dosbox-x`. previously it failed with a timeout error.

if this is what you wanted, I will make the change.

Reply:
Yep, that's pretty much what I had in mind. Feel free to commit that change!

I'm not sure how familiar you are with the concept of basic blocks (see the Wikipedia article about it [here](https://en.wikipedia.org/wiki/Basic_block)), so I won't ask you to do anything else. The fact that we're not timing out anymore on that project is great!

If you want to rewrite it so that you have the same behavior as before, you need to slightly change your rewrite. Basically, to get the behavior of `e.getASuccessor*() = e2` while retaining good performance you need to rewrite it to 2 cases:
1. the first case is almost like you said: `e1.getBasicBlock().getASuccessor+() = e2.getBasicBlock()`.
  
Note the use of `+` instead of `*`. This handles the case where `e1` and `e2` are in different basic blocks. You should not use `*` here since that would allow for `e1` and `e2` to be in the same block, and `e2` to come earlier in the basic block than `e1`.

2. the second case is then handles the opposite condition: when `e1` and `e2` are in the _same_ basic block, and `e2` comes after `e1` in that block. This should be handles by a case like:
```ql
exists(BasicBlock bb, int i1, int i2 | 
  e1.getBasicBlock() = bb and
  e2.getBasicBlock() = bb and
  bb.getNode(i1) = e1 and
  bb.getNode(i2) = e2 and
  i1 < i2
)

```

Reply:
Thanks again for the detailed clarifications.
I am very interested in increasing the depth of my knowledge and the quality of the code.
I did not immediately use this approach, since I still have not figured out the tool for measuring the speed of a request. (as mentioned earlier)

your suggestion to use `*` `+`.
I understood the principle, but in this request, the relative position is not important, only the reachability is important. so I made changes.